packages/expo-file-system/ios/EXFileSystem/EXFileSystem.m

```diff
@@ -556,46 +544,91 @@ - (NSDictionary *)encodingMap
+  EXSessionTaskDelegate *exTask = [[EXSessionDownloadTaskDelegate alloc] initWithResolve:resolve
+                                                                                  reject:reject
+                                                                                localUrl:localUri
+                                                                      shouldCalculateMd5:[options[@"md5"]boolValue]];
+  [self.tasks setObject:exTask forKey:task];
+  [task resume];
+}
+
+UM_EXPORT_METHOD_AS(uploadAsync,
+                    uploadAsync:(NSString *)urlString
+                       localURI:(NSString *)fileUriString
+                        options:(NSDictionary *)options
+                       resolver:(UMPromiseResolveBlock)resolve
+                       rejecter:(UMPromiseRejectBlock)reject)
+{
+  NSURL *fileUri = [NSURL URLWithString:fileUriString];
+  if (![fileUri.scheme isEqualToString:@"file"]) {
+    reject(@"E_FILESYSTEM_PERMISSIONS",
+           [NSString stringWithFormat:@"Cannot upload file '%@'. Only 'file://' URI are supported.", fileUri],
+           nil);
+    return;
   }
-  sessionConfiguration.URLCache = nil;
-  NSURLSession *session = [NSURLSession sessionWithConfiguration:sessionConfiguration];
-  NSURLSessionDataTask *task = [session dataTaskWithURL:url completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
-    if (error) {
-      reject(@"E_DOWNLOAD_FAILED",
-             [NSString stringWithFormat:@"Could not download from '%@'", url],
-             error);
+  if (!([self _checkIfFileExists:fileUri.path])) {
+    reject(@"E_FILE_NOT_EXISTS",
+           [NSString stringWithFormat:@"File '%@' does not exist.", fileUri],
+           nil);
+    return;
+  }
+  if (![self _checkHeadersDictionary:options[@"headers"]]) {
+    reject(@"ERR_FILESYSTEM_INVALID_HEADERS_DICTIONARY",
+           @"Invalid headers dictionary. Keys and values should be strings.",
+           nil);
+    return;
+  }
+
+  NSMutableURLRequest *urlRequest = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:urlString]];
+  if (options[@"httpMethod"]) {
+    NSString *httpMethod = [self _importHttpMethod:options[@"httpMethod"]];
+    if (!httpMethod) {
+      reject(@"ERR_FILESYSTEM_INVALID_HTTP_METHOD",
+             [NSString stringWithFormat:@"Invalid http method %@", options[@"httpMethod"]],
+             nil);
       return;
     }
-    [data writeToFile:path atomically:YES];
-    
-    NSHTTPURLResponse *httpResponse = (NSHTTPURLResponse *)response;
-    NSMutableDictionary *result = [NSMutableDictionary dictionary];
-    result[@"uri"] = [NSURL fileURLWithPath:path].absoluteString;
-    if (options[@"md5"]) {
-      result[@"md5"] = [data md5String];
+    [urlRequest setHTTPMethod:httpMethod];
+  }
+  NSDictionary *headers = options[@"headers"];
+  if (headers != nil) {
+    for (NSString *headerKey in headers) {
+        [urlRequest setValue:[headers valueForKey:headerKey] forHTTPHeaderField:headerKey];
```

Comment:
Could you please extract the `NSMutableURLRequest` creation to a separate function? This way we will have the headers-setting code only in one place!